langstream-runtime/langstream-runtime-impl/src/main/java/ai/langstream/runtime/agent/AgentCodeDownloader.java

```diff
@@ -16,62 +16,69 @@
+        log.info("Downloading custom code {}, cluster configuration {}", configuration, clusterConfiguration);
+        final String tenant = Objects.requireNonNull(configuration.tenant());
+        final String codeDownloadPath = Objects.requireNonNull(configuration.codeDownloadPath());
+        final String applicationId = Objects.requireNonNull(configuration.applicationId());
+        final String codeArchiveId = Objects.requireNonNull(configuration.codeArchiveId());
+        final Path destinationPath = Paths.get(codeDownloadPath);
+        if (destinationPath.toFile().exists()) {
+            destinationPath.toFile().mkdirs();
+        }
 
-            Path codeConfigPath = Path.of(args[0]);
-            log.info("Loading code config config from {}", codeConfigPath);
-            final CodeStorageConfig codeStorageConfig =
-                    MAPPER.readValue(codeConfigPath.toFile(), CodeStorageConfig.class);
-            Path codeDownloadPath = Path.of(args[1]);
-            downloadCustomCode(codeStorageConfig, codeDownloadPath);
-            log.info("Code downloaded to {}", codeDownloadPath);
-        } catch (Throwable error) {
-            errorHandler.handleError(error);
-            return;
+        try (AdminClient adminClient = createAdminClient(clusterConfiguration, token, tenant)) {
+            final HttpResponse<InputStream> download = adminClient.applications()
+                    .download(applicationId, codeArchiveId, HttpResponse.BodyHandlers.ofInputStream());
+            LocalZipFileArchiveFile.extractTo(download.body(), destinationPath);
+            log.info("Downloaded code to {}", codeDownloadPath);
         }
     }
 
-    private static void downloadCustomCode(CodeStorageConfig codeStorageConfig, Path codeDownloadPath) throws Exception {
+    private static AdminClient createAdminClient(ClusterConfiguration clusterConfiguration, String token, String tenant) throws Exception {
+        final ai.langstream.admin.client.AdminClientConfiguration config =
+                ai.langstream.admin.client.AdminClientConfiguration.builder()
+                        .webServiceUrl(clusterConfiguration.controlPlaneUrl())
+                        .token(token)
+                        .tenant(tenant)
+                        .build();
 
+        final AdminClientLogger logger = new AdminClientLogger() {
+            @Override
+            public void log(Object message) {
+                log.info(message.toString());
+            }
 
+            @Override
+            public void error(Object message) {
+                log.error(message.toString());
 
-        if (codeStorageConfig != null) {
-            log.info("Downloading custom code from {}", codeStorageConfig);
-            log.info("Custom code is stored in {}", codeDownloadPath);
-            try (CodeStorage codeStorage =
```

Comment:
I wonder if we could have kept this new code as simply a new implenetation of CodeStorage.
ControlPlaneCodeStorage,
now we are binding strictly the download of the code to the control plane and we don't have anymore this abstraction here.

Reply:
I thought about that but the problem it's that the control plane download is not "generic". You need to read a secret for the control plane url, another secret for the token. This impacts both the deployer and the runtime. In the deployer you'd need to ask the code storage impl to customize the pod or load secrets or similar. It doesn't feel very generic. 

Additionally, this is generally more secure since the we know for sure that the control plane url is verified.

Another advantage is that you need to write the code storage configuration only in the control plane now, which facilitates upgrades